pkg/eventemitter/eventemitter.go

```diff
@@ -0,0 +1,356 @@
+			log.Error(err, "Failed to update CloudEventSourceStatus")
+		}
+	}
+}
+
+// Emit is emitting event to both local kubernetes and custom CloudEventSource handler. After emit event to local kubernetes, event will inqueue and waitng for handler's consuming.
+func (e *EventEmitter) Emit(object runtime.Object, namesapce types.NamespacedName, eventtype, reason, message string) {
+	e.EventRecorder.Event(object, eventtype, reason, message)
+
+	if len(e.eventHandlersCache) == 0 {
+		return
+	}
+
+	name, _ := meta.NewAccessor().Name(object)
+	eventData := EventData{
+		namespace:  namesapce.Namespace,
+		objectName: name,
+		eventtype:  eventtype,
+		reason:     reason,
+		message:    message,
+		time:       time.Now().UTC(),
+	}
+	go e.enqueueEventData(eventData)
+}
+
+func (e *EventEmitter) enqueueEventData(eventData EventData) {
+	select {
+	case ch <- eventData:
+		log.V(1).Info("Event enqueued successfully.")
+	case <-time.After(MaxWaitingEnqueueTime * time.Second):
+		log.Error(nil, "Event cannot enqueue. Need to be check if handler can emit events.")
+	}
+}
+
+// emitEventByHandler handles event emitting. It will follow these logic:
+// 1. If there is a new EventData, call all handlers for emitting.
+// 2. Once there is an error when emitting event, record the handler's key and reqeueu this EventData.
+// 3. If the maximum number of retries has been exceeded, discard this event.
+func (e *EventEmitter) emitEventByHandler(eventData EventData) {
+	if eventData.retryTimes >= MaxRetryTimes {
+		log.Error(eventData.err, "Failed to emit Event multiple times. Will drop this event and need to check if event endpoint works well", "CloudEventSource", eventData.objectName)
+		e.emitErrorHandle(eventData, eventData.err)
+		return
+	}
+
+	if eventData.handlerKey == "" {
+		for key, handler := range e.eventHandlersCache {
+			eventData.handlerKey = key
+			if handler.GetActiveStatus() == metav1.ConditionTrue {
+				go handler.EmitEvent(eventData, e.emitErrorHandle)
+			} else {
+				log.V(1).Info("EventHandler's status is not active. Please check if event endpoint works well", "CloudEventSource", eventData.objectName)
+			}
+		}
+	} else {
+		log.Info("Reemit event failed", "handler", eventData.handlerKey, "retry times", eventData.retryTimes)
+		handler, found := e.eventHandlersCache[eventData.handlerKey]
+		if found && handler.GetActiveStatus() == metav1.ConditionTrue {
+			go handler.EmitEvent(eventData, e.emitErrorHandle)
+		}
+	}
+}
+
+func (e *EventEmitter) emitErrorHandle(eventData EventData, err error) {
+	if eventData.retryTimes >= MaxRetryTimes {
+		log.V(1).Info("Failed to emit Event multiple times. Will set handler failure status.", "handler", eventData.handlerKey, "retry times", eventData.retryTimes)
+		handler, found := e.eventHandlersCache[eventData.handlerKey]
+		if found {
+			handler.SetActiveStatus(metav1.ConditionFalse)
```

Comment:
Does any part of the code ever set this back to true?

From what I am able to tell, once the event handler fails to send a particular cloud event 5 times, user will need to restart keda-operator for the handler to be again operational. Am I missing something?

Reply:
No. 
Users can recreate the CloudEventSource to resend the event instead of restarting the KEDA operator. We could consider adding an annotation to the CloudEventSource to control this behavior. I think it can be in next PR. What do you think? @tomkerkhove

Reply:
ok, good to know. I would be in favour of not closing https://github.com/kedacore/keda/issues/3533 quite yet and we can mention all the follow-up tasks there including this one and track the progress independently. Recreation of `CloudEventSource` didn't cross my mind and while not ideal, it's much better than restarting `keda-operator` :)

Reply:
* I agree we that after this PR is merged we should think about retry/backoff functionality 
* Current status functionality doesn't seem to work - I always see the default condition `Active: Unknown` (even after sending messages) - I created a [PR](https://github.com/SpiritZhou/keda/pull/1/files#diff-86c8d5f0a95aab93aa2043c700815c4df67295d4bc4e3430583153a7c367757bR354) with a fix against @SpiritZhou repository, can you take a look?

Reply:
@josefkarasek Thanks! I think similar improvements such as refactoring the init() function could also be made in other controllers. 👍